README
NAME

    WWW::ProxyChecker - check whether or not proxy servers are alive

SYNOPSIS

        use strict;
        use warnings;
        use WWW::ProxyChecker;
    
        my $checker = WWW::ProxyChecker->new( debug => 1 );
    
        my $working_ref= $checker->check( [ qw(
                    http://221.139.50.83:80
                    http://111.111.12.83:8080
                    http://111.111.12.183:3218
                    http://111.111.12.93:8080
                )
            ]
        );
    
        die "No working proxies were found\n"
            if not @$working_ref;
    
        print "$_ is alive\n"
            for @$working_ref;

DESCRIPTION

    The module provides means to check whether or not HTTP proxies are
    alive. The module was designed more towards "quickly scanning through
    to get a few" than "guaranteed or your money back" therefore there is
    no 100% guarantee that non-working proxies are actually dead and that
    all of the reported working proxies are actually good.

CONSTRUCTOR

 new

        my $checker = WWW::ProxyChecker->new;
    
        my $checker_juicy = WWW::ProxyChecker->new(
            timeout       => 5,
            max_kids      => 20,
            max_working_per_child => 2,
            check_sites   => [ qw(
                    http://google.com
                    http://microsoft.com
                    http://yahoo.com
                    http://digg.com
                    http://facebook.com
                    http://myspace.com
                )
            ],
            agent   => 'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.8.1.12)'
                        .' Gecko/20080207 Ubuntu/7.10 (gutsy) Firefox/2.0.0.12',
            debug => 1,
        );

    Bakes up and returns a new WWW::ProxyChecker object. Takes a few
    arguments all of which are optional. Possible arguments are as follows:

  timeout

        ->new( timeout => 5 );

    Optional. Specifies timeout in seconds to give to LWP::UserAgent object
    which is used for checking. If a connection to the proxy times out the
    proxy is considered dead. The lower the value, the faster the check
    will be done but also the more are the chances that you will throw away
    good proxies. Defaults to: 5 seconds

  agent

        ->new( agent => 'ProxeyCheckerz' );

    Optional. Specifies the User Agent string to use while checking
    proxies. By default will be set to mimic Firefox.

  check_sites

        ->new( check_sites => [ qw( http://some_site.com http://other.com ) ] );

    Optional. Takes an arrayref of sites to try to connect to through a
    proxy. Yes! It's evil, saner ideas are more than welcome. Defaults to:

        check_sites   => [ qw(
                    http://google.com
                    http://microsoft.com
                    http://yahoo.com
                    http://digg.com
                    http://facebook.com
                    http://myspace.com
                )
            ],

  max_kids

        ->new( max_kids => 20 );

    Optional. Takes a positive integer as a value. The module will fork up
    maximum of max_kids processes to check proxies simultaneously. It will
    fork less if the total number of proxies to check is less than
    max_kids. Technically, setting this to a higher value might speed up
    the overall process but keep in mind that it's the number of
    simultaneous connections that you will have open. Defaults to: 20

  max_working_per_child

        ->new( max_working_per_child => 2 );

    Optional. Takes a positive integer as a value. Specifies how many
    working proxies each sub process should find before aborting (it will
    also abort if proxy list is exhausted). In other words, setting 20
    max_kids and max_working_per_child to 2 will give you 40 working
    proxies at most, no matter how many are in the original list.
    Specifying undef will get rid of limit and make each kid go over the
    entire sub list it was given. Defaults to: undef (go over entire sub
    list)

  debug

        ->new( debug => 1 );

    Optional. When set to a true value will make the module print out some
    debugging information (which proxies failed and how, etc). By default
    not specifies (debug is off)

METHODS

 check

        my $working_ref = $checker->check( [ qw(
                    http://221.139.50.83:80
                    http://111.111.12.83:8080
                    http://111.111.12.183:3218
                    http://111.111.12.93:8080
                )
            ]
        );

    Instructs the object to check several proxies. Returns a (possibly
    empty) array ref of addresses which the object considers to be alive
    and working. Takes an arrayref of proxy addresses. The elements of this
    arrayref will be passed to LWP::UserAgent's proxy() method as:

        $ua->proxy( [ 'http', 'https', 'ftp', 'ftps' ], $proxy );

    so you can read the docs for LWP::UserAgent and maybe think up
    something creative.

 alive

        my $last_alive = $checker->alive;

    Must be called after a call to check(). Takes no arguments, returns the
    same arrayref last check() returned.

ACCESSORS/MUTATORS

    The module provides an accessor/mutator for each of the arguments in
    the constructor (new() method). Calling any of these with an argument
    will set a new value. All of these return a currently set value:

        max_kids
        check_sites
        max_working_per_kid
        timeout
        agent
        debug

    See CONSTRUCTOR section for more information about these.

REPOSITORY

    https://github.com/stevieb9/p5-www-proxychecker

BUGS

    To report bugs or request features, please use
    https://github.com/stevieb9/p5-www-proxychecker/issues

AUTHOR

    Zoffix Znet, <zoffix at cpan.org> (http://zoffix.com,
    http://haslayout.net)

    Adopted on Feb 4, 2016 and currently maintained by:

    Steve Bertrand <steveb at cpan.org>

COPYRIGHT & LICENSE

    Copyright 2016 Steve Bertrand

    Copyright 2008 Zoffix Znet, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for WWW-ProxyChecker

1.005   xxx
        - added fastest(), returns array of successful proxies ordered
          from fastest to slowest
        - remove perl 5.8 from travis

1.004   2016-02-07
        - fixed MANIFEST
        - added t/02-check.t

1.003   2016-02-07
        - distribution adopted by Steve Bertrand (STEVEB) on
          2016-02-04
        - removed distzilla
        - updated POD

1.001001 2014-03-13
        - Maintenance update to fix up failing tests due to
          outdated code usage
        - Converted to dzil

0.002   2008-03-16
        - fixed the documentation which was missing
          'max_working_per_kid' description

0.001   2008-03-16
        - First version, released on an unsuspecting world.



Makefile.PL
use strict;
use warnings;
 
use 5.006;
 
use ExtUtils::MakeMaker 6.30;
 
my %WriteMakefileArgs = (
    "ABSTRACT" => "check whether or not proxy servers are alive",
    "AUTHOR" => "Zoffix Znet <cpan\@zoffix.com>",
    "BUILD_REQUIRES" => {},
    "CONFIGURE_REQUIRES" => {
        "ExtUtils::MakeMaker" => "6.30"
    },
    "DISTNAME" => "WWW-ProxyChecker",
    "EXE_FILES" => [],
    "LICENSE" => "perl",
    "NAME" => "WWW::ProxyChecker",
    "META_MERGE" => {
        'meta-spec' => { version => 2 },
        resources => {
            bugtracker => {
                web => 'https://github.com/stevieb9/p5-www-proxychecker/issues',
            },
            repository => {
                type => 'git',
                url => 'https://github.com/stevieb9/p5-www-proxychecker.git',
                web => 'https://github.com/stevieb9/p5-www-proxychecker',
            },
        },
    },

    "PREREQ_PM" => {
        "Carp" => 0,
        "Class::Accessor::Grouped" => 0,
        "IO::Pipe" => 0,
        "LWP::UserAgent" => 0,
        "base" => 0,
        "strict" => 0,
        "warnings" => 0
    },
    "TEST_REQUIRES" => {
        "File::Spec" => 0,
        "IO::Handle" => 0,
        "IPC::Open3" => 0,
        "Test::More" => 0
    },
    "VERSION_FROM" => 'lib/WWW/ProxyChecker.pm',
    "test" => {
        "TESTS" => "t/*.t"
    }
);


my %FallbackPrereqs = (
    "Carp" => 0,
    "Class::Accessor::Grouped" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IO::Pipe" => 0,
    "IPC::Open3" => 0,
    "LWP::UserAgent" => 0,
    "Test::More" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


examples/check.pl
use strict;
use warnings;

use WWW::ProxyChecker;

my $checker = WWW::ProxyChecker->new( debug => 1 );

for ( @{ $checker->check( _make_data() ) } ) {
    print "$_ is alive\n";
}

sub _make_data {
    my $prox = [
            {
              'country' => 'South Korea',
              'ip' => '221.139.50.83',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Czechoslovakia',
              'ip' => '82.114.70.98',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'China',
              'ip' => '202.116.76.163',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '80'
            },
            {
              'country' => 'N/A',
              'ip' => '190.78.95.118',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Argentina',
              'ip' => '200.51.41.29',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Australia',
              'ip' => '165.228.132.10',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'China',
              'ip' => '211.161.128.126',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Brazil',
              'ip' => '200.179.148.141',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'N/A',
              'ip' => '189.33.153.221',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Romania',
              'ip' => '81.181.45.25',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Japan',
              'ip' => '203.143.116.101',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '80'
            },
            {
              'country' => 'Mexico',
              'ip' => '189.131.235.179',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '80'
            },
            {
              'country' => 'South Korea',
              'ip' => '211.119.242.47',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '80'
            },
            {
              'country' => 'Brazil',
              'ip' => '200.171.124.197',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Ghana',
              'ip' => '41.204.32.35',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'China',
              'ip' => '219.232.224.35',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'N/A',
              'ip' => '78.39.204.114',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '8080'
            },
            {
              'country' => 'Egypt',
              'ip' => '84.205.98.242',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '8080'
            },
            {
              'country' => 'Romania',
              'ip' => '85.186.159.233',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Iran',
              'ip' => '85.9.83.51',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Great Britain (UK)',
              'ip' => '195.248.254.11',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Indonesia',
              'ip' => '203.190.52.115',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'N/A',
              'ip' => '222.127.228.19',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '80'
            },
            {
              'country' => 'Argentina',
              'ip' => '190.3.10.228',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'South Africa',
              'ip' => '196.30.229.162',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '80'
            },
            {
              'country' => 'Ecuador',
              'ip' => '200.125.201.218',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Brazil',
              'ip' => '200.219.152.9',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'China',
              'ip' => '218.59.164.246',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Brazil',
              'ip' => '143.107.238.188',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'United States',
              'ip' => '209.129.192.52',
              'last_test' => '2008-03-14',
              'type' => 'high anonymity',
              'port' => '80'
            },
            {
              'country' => 'Slovak Republic',
              'ip' => '213.160.169.244',
              'last_test' => '2008-03-14',
              'type' => 'high anonymity',
              'port' => '8080'
            },
            {
              'country' => 'United Arab Emirates',
              'ip' => '195.229.236.106',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '80'
            },
            {
              'country' => 'United States',
              'ip' => '65.213.194.11',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '80'
            },
            {
              'country' => 'Turkey',
              'ip' => '212.58.11.72',
              'last_test' => '2008-03-14',
              'type' => 'anonymous',
              'port' => '80'
            },
            {
              'country' => 'Japan',
              'ip' => '219.66.154.224',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Indonesia',
              'ip' => '202.146.129.220',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Great Britain (UK)',
              'ip' => '217.41.27.254',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Thailand',
              'ip' => '203.149.32.4',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Brazil',
              'ip' => '200.219.152.6',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Philippines',
              'ip' => '124.104.66.27',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Japan',
              'ip' => '202.216.177.18',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Brazil',
              'ip' => '200.179.148.130',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            },
            {
              'country' => 'Thailand',
              'ip' => '202.29.87.39',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Australia',
              'ip' => '165.228.131.10',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Australia',
              'ip' => '165.228.128.10',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'China',
              'ip' => '218.87.71.138',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'United States',
              'ip' => '63.149.98.82',
              'last_test' => '2008-03-14',
              'type' => 'high anonymity',
              'port' => '80'
            },
            {
              'country' => 'India',
              'ip' => '203.110.81.218',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'Brazil',
              'ip' => '200.242.179.51',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '3128'
            },
            {
              'country' => 'N/A',
              'ip' => '208.72.247.201',
              'last_test' => '2008-03-14',
              'type' => 'high anonymity',
              'port' => '3128'
            },
            {
              'country' => 'United States',
              'ip' => '63.149.98.27',
              'last_test' => '2008-03-14',
              'type' => 'high anonymity',
              'port' => '80'
            },
            {
              'country' => 'Indonesia',
              'ip' => '202.173.23.141',
              'last_test' => '2008-03-14',
              'type' => 'transparent',
              'port' => '8080'
            }
    ];

    my @data;
    for ( @$prox ) {
        push @data, join '', 'http://', join ':', @$_{qw(ip port)};
    }
    return \@data;
}

lib/WWW/ProxyChecker.pm
package WWW::ProxyChecker;

use warnings;
use strict;

# VERSION

our $VERSION = '1.006';

use Data::Dumper;
use Carp;
use LWP::UserAgent;
use IO::Pipe;
use Time::HiRes qw(gettimeofday);

use base 'Class::Accessor::Grouped';
__PACKAGE__->mk_group_accessors(simple => qw/
    max_kids
    debug
    alive
    check_sites
    max_working_per_kid
    timeout
    agent
/);

sub new {
    my $self = bless {}, shift;
    croak "Must have even number of arguments to new()"
        if @_ & 1;

    my %args = @_;
    $args{ +lc } = delete $args{ $_ } for keys %args;

    %args = (
        timeout       => 5,
        max_kids      => 20,
        check_sites   => [ qw(
                http://google.com
                http://microsoft.com
                http://yahoo.com
                http://digg.com
                http://facebook.com
                http://myspace.com
            )
        ],
        agent   => 'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.8.1.12)'
                    .' Gecko/20080207 Ubuntu/7.10 (gutsy) Firefox/2.0.0.12',

        %args,
    );

    $self->$_( $args{ $_ } ) for keys %args;

    return $self;
}
sub check {
    my ( $self, $proxies_ref ) = @_;

    $self->alive(undef);
    $self->{fastest} = {};

    warn "About to check " . @$proxies_ref . " proxies\n"
        if $self->debug;

    my $working_ref = $self->_start_checker( @$proxies_ref );

    warn @$working_ref . ' out of ' . @$proxies_ref
            . " seem to be alive\n" if $self->debug;

    return $self->alive($working_ref);
}
sub fastest {
    my $self = shift;

    my @debug_list;
    my @fastest;

    my %list = %{ $self->{fastest} };

    for my $proxy (sort { $list{$a} <=> $list{$b} } keys %list){
        $list{$proxy} = sprintf("%.2f", $list{$proxy});
        push @debug_list, "$proxy :: $list{$proxy}";
        push @fastest, $proxy;
    }

    if ($self->debug) {
        warn "$_\n" for @debug_list;
    }

    return \@fastest;
}
sub _start_checker {
    my ( $self, @proxies ) = @_;

    my $n = $self->max_kids;
    $n > @proxies and $n = @proxies;
    my $mod = @proxies / $n;
    my %prox;
    for ( 1 .. $n ) {
        $prox{ $_ } = [ splice @proxies, 0,$mod ]
    }
    push @{ $prox{ $n } }, @proxies; # append any left over addresses

    $SIG{CHLD} = 'IGNORE';
    my @children;
    for my $num ( 1 .. $self->max_kids ) {
        my $pipe = new IO::Pipe;
        my $time;

        if ( my $pid = fork ) { # parent
            $pipe->reader;
            push @children, $pipe;
        }
        elsif ( defined $pid ) { # kid
            $pipe->writer;

            my $ua = LWP::UserAgent->new(
                timeout => $self->timeout,
                agent   => $self->agent,
            );

            my $check_sites_ref = $self->check_sites;
            my $debug = $self->debug;
            my @working;
            for my $proxy ( @{ $prox{ $num } } ) {
                warn "Checking $proxy in kid $num\n"
                    if $debug;

                if ($time = $self->_check_proxy($ua, $proxy, $check_sites_ref) ) {
                    push @working, $proxy;

                    last
                        if defined $self->max_working_per_kid
                            and @working >= $self->max_working_per_kid;
                }
            }
            print $pipe "$_ $time\n" for @working;
            exit;
        }
        else { # error
            carp "Failed to fork kid number $num ($?)";
        }

    }

    my @working_proxies;
    for my $num ( 0 .. $#children ) {
        my $fh = $children[$num];
        while (<$fh>) {
            chomp;
            my ($proxy, $time) = split;
            $self->{fastest}{$proxy} = $time if $time != 1;
            push @working_proxies, $proxy;
        }
    }

    return \@working_proxies;
}
sub _check_proxy {
    my ( $self, $ua, $proxy, $sites_ref ) = @_;

    $ua->proxy( [ 'http', 'https', 'ftp', 'ftps' ], $proxy);

    my $start = gettimeofday();

    my $response = $ua->get( $sites_ref->[rand @$sites_ref] );

    my $time = gettimeofday() - $start;

    if ( $response->is_success ){
        return $time;
    }
    else {
        warn "Failed on $proxy " . $response->status_line . "\n"
            if $self->debug;

        my $response_code = $response->code;
        return 0
            if grep { $response_code eq $_ } qw(407 502 503 403);

        ( my $proxy_no_scheme = $proxy ) =~ s{(?:ht|f)tps?://}{}i;
        return $response->status_line
        =~ /^500 read timeout$|\Q$proxy_no_scheme/ ? 0 : 1;
    }
}

1;
__END__

=head1 NAME

WWW::ProxyChecker - Check whether or not proxy servers are alive

=for html
<a href="http://travis-ci.org/stevieb9/p5-www-proxychecker"><img src="https://secure.travis-ci.org/stevieb9/p5-www-proxychecker.png"/>
<a href='https://coveralls.io/github/stevieb9/p5-www-proxychecker?branch=master'><img src='https://coveralls.io/repos/stevieb9/p5-www-proxychecker/badge.svg?branch=master&service=github' alt='Coverage Status' /></a>

=head1 SYNOPSIS

    use strict;
    use warnings;

    use WWW::ProxyChecker;

    my $checker = WWW::ProxyChecker->new(debug => 1);

    my $working_ref = $checker->check( [ qw(
                http://221.139.50.83:80
                http://111.111.12.83:8080
                http://111.111.12.183:3218
                http://111.111.12.93:8080
            )
        ]
    );

    die "No working proxies were found\n"
        if not @$working_ref;

    print "$_ is alive\n"
        for @$working_ref;

    my $fastest = $checker->fastest;

    print "fastest to slowest: $_\n" for @$fastest;

    # Used in conjunction with WWW::FreeProxyListsCom, you can easily
    # find the fastest available proxies automatically

    use WWW::FreeProxyListsCom;

    my $proxy_fetcher = WWW::FreeProxyListsCom->new(timeout => 10);
    my $checker = WWW::ProxyChecker->new;

    $proxy_fetcher->get_list(type => 'us');

    $fast_to_slow = $proxy_fetcher->filter(is_https => 'true');

    print "$_\n" for @$fast_to_slow;

=head1 DESCRIPTION
 
The module provides means to check whether or not HTTP proxies are alive. It
was designed more towards "quickly scanning through to get a few"
than "guaranteed or your money back" therefore there is no 100% guarantee
that non-working proxies are actually dead and that all of the reported
working proxies are actually good.

Using this module in conjunction, you can automate finding available proxies,
even sorting the available ones fastest to slowest.

=head1 CONSTRUCTOR

=head2 new

    my $checker = WWW::ProxyChecker->new;

    my $checker_juicy = WWW::ProxyChecker->new(
        timeout       => 5,
        max_kids      => 20,
        max_working_per_child => 2,
        check_sites   => [ qw(
                http://google.com
                http://microsoft.com
                http://yahoo.com
                http://digg.com
                http://facebook.com
                http://myspace.com
            )
        ],
        agent   => 'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.8.1.12)'
                    .' Gecko/20080207 Ubuntu/7.10 (gutsy) Firefox/2.0.0.12',
        debug => 1,
    );

Bakes up and returns a new WWW::ProxyChecker object. Takes a few arguments
I<all of which are optional>. Possible arguments are as follows:

=head3 timeout

    ->new( timeout => 5 );

B<Optional>. Specifies timeout in seconds to give to L<LWP::UserAgent>
object which
is used for checking. If a connection to the proxy times out the proxy
is considered dead. The lower the value, the faster the check will be
done but also the more are the chances that you will throw away good
proxies. B<Defaults to:> C<5> seconds

=head3 agent

    ->new( agent => 'ProxeyCheckerz' );

B<Optional>. Specifies the User Agent string to use while checking proxies. B<By default> will be set to mimic Firefox.

=head3 check_sites

    ->new( check_sites => [ qw( http://some_site.com http://other.com ) ] );

B<Optional>. Takes an arrayref of sites to try to connect to through a
proxy. Yes! It's evil, saner ideas are more than welcome. B<Defaults to:>

    check_sites   => [ qw(
                http://google.com
                http://microsoft.com
                http://yahoo.com
                http://digg.com
                http://facebook.com
                http://myspace.com
            )
        ],

=head3 max_kids

    ->new( max_kids => 20 );

B<Optional>. Takes a positive integer as a value.
The module will fork up maximum of C<max_kids> processes to check proxies
simultaneously. It will fork less if the total number of proxies to check
is less than C<max_kids>. Technically, setting this to a higher value
might speed up the overall process but keep in mind that it's the number
of simultaneous connections that you will have open. B<Defaults to:> C<20>

=head3 max_working_per_child

    ->new( max_working_per_child => 2 );

B<Optional>. Takes a positive integer as a value. Specifies how many
working proxies each sub process should find before aborting (it will
also abort if proxy list is exhausted). In other words, setting C<20>
C<max_kids> and C<max_working_per_child> to C<2> will give you 40 working
proxies at most, no matter how many are in the original list. Specifying
C<undef> will get rid of limit and make each kid go over the entire sub
list it was given. B<Defaults to:> C<undef> (go over entire sub list)

=head3 debug

    ->new( debug => 1 );

B<Optional>. When set to a true value will make the module print out
some debugging information (which proxies failed and how, etc).
B<By default> not specifies (debug is off)

=head1 METHODS

=head2 check

    my $working_ref = $checker->check( [ qw(
                http://221.139.50.83:80
                http://111.111.12.83:8080
                http://111.111.12.183:3218
                http://111.111.12.93:8080
            )
        ]
    );

Instructs the object to check several proxies. Returns a (possibly empty)
array ref of addresses which the object considers to be alive and working.
Takes an arrayref of proxy addresses. The elements of this arrayref will
be passed to L<LWP::UserAgent>'s C<proxy()> method as:

    $ua->proxy( [ 'http', 'https', 'ftp', 'ftps' ], $proxy );

so you can read the docs for L<LWP::UserAgent> and maybe think up something
creative.

=head2 alive

    my $last_alive = $checker->alive;

Must be called after a call to C<check()>. Takes no arguments, returns
the same arrayref last C<check()> returned.

=head2 fastest

    my $fastest = $checker->fastest;

Returns an array reference of alive proxies, ordered by fastest return time
to slowest.

=head1 ACCESSORS/MUTATORS

The module provides an accessor/mutator for each of the arguments in
the constructor (new() method). Calling any of these with an argument
will set a new value. All of these return a currently set value:

    max_kids
    check_sites
    max_working_per_kid
    timeout
    agent
    debug

See C<CONSTRUCTOR> section for more information about these.

=head1 REPOSITORY

L<https://github.com/stevieb9/p5-www-proxychecker>

=head1 BUGS

To report bugs or request features, please use
L<https://github.com/stevieb9/p5-www-proxychecker/issues>

=head1 AUTHOR

Zoffix Znet, C<< <zoffix at cpan.org> >>
(L<http://zoffix.com>, L<http://haslayout.net>)

Adopted on Feb 4, 2016 and currently maintained by:

Steve Bertrand C<< <steveb at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2016,2018 Steve Bertrand

Copyright 2008 Zoffix Znet, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.


=cut
